src/ui/SmartSnippet/SmartSnippet.ts

```diff
@@ -0,0 +1,140 @@
+  HAS_ANSWER_CLASSNAME,
+  SHADOW_CLASSNAME,
+  CONTENT_CLASSNAME,
+  SOURCE_CLASSNAME,
+  SOURCE_TITLE_CLASSNAME,
+  SOURCE_URL_CLASSNAME
+};
+
+export interface ISmartSnippetOptions {}
+
+export class SmartSnippet extends Component {
+  static ID = 'SmartSnippet';
+
+  static options: ISmartSnippetOptions = {};
+
+  static doExport = () => {
+    exportGlobally({
+      SmartSnippet
+    });
+  };
+
+  private shadowContainer: HTMLElement;
+  private sourceContainer: HTMLElement;
+  private snippetContainer: HTMLElement;
+
+  constructor(public element: HTMLElement, public options?: ISmartSnippetOptions, bindings?: IComponentBindings) {
+    super(element, SmartSnippet.ID, bindings);
+    this.bind.onRootElement(QueryEvents.querySuccess, (data: IQuerySuccessEventArgs) => this.handleQuerySuccess(data));
+  }
+
+  private get style() {
+    return $$(this.element)
+      .children()
+      .filter(element => element instanceof HTMLScriptElement && element.type.toLowerCase() === 'text/css')
+      .map(element => element.innerHTML)
+      .join('\n');
+  }
+
+  private set hasAnswer(hasAnswer: boolean) {
+    this.element.classList.toggle(HAS_ANSWER_CLASSNAME, hasAnswer);
+  }
+
+  public createDom() {
+    this.buildShadow();
+    this.buildSourceContainer();
+  }
+
+  private buildShadow() {
+    this.element.appendChild((this.shadowContainer = $$('div', { className: SHADOW_CLASSNAME }).el));
+    const shadow = this.shadowContainer.attachShadow({ mode: 'open' });
+    shadow.appendChild((this.snippetContainer = $$('section', { className: CONTENT_CLASSNAME }).el));
+    shadow.appendChild(this.buildStyle());
+  }
+
+  private buildSourceContainer() {
+    this.element.appendChild((this.sourceContainer = $$('div', { className: SOURCE_CLASSNAME }).el));
+  }
+
+  private buildStyle() {
+    const style = this.style;
+    if (!style) {
+      return;
+    }
+    const styleTag = document.createElement('style');
+    styleTag.innerHTML = style;
+    return styleTag;
+  }
+
+  /**
+   * @warning This method only works for the demo. In practice, the source of the answer will not always be part of the results.
+   */
+  private getCorrespondingResult(questionAnswer: IQueryAnswerResponse) {
+    return find(
+      this.queryController.getLastResults().results,
+      result => result.raw[questionAnswer.documentId.contentIdKey] === questionAnswer.documentId.contentIdValue
+    );
+  }
+
+  private handleQuerySuccess(data: IQuerySuccessEventArgs) {
+    const { questionAnswer } = data.results;
+    if (!questionAnswer) {
+      this.hasAnswer = false;
+      return;
+    }
+    this.hasAnswer = true;
+    this.render(questionAnswer);
+  }
+
+  private render(questionAnswer: IQueryAnswerResponse) {
+    this.ensureDom();
+    this.renderSnippet(questionAnswer.answerSnippet);
+    const correspondingResult = this.getCorrespondingResult(questionAnswer);
+    if (!correspondingResult) {
+      return;
+    }
+    this.renderSource(correspondingResult);
+  }
```

Comment:
It would be good to add some unit tests, especially for the functionality of the render method checking that the snippet is added correctly along with the source result.

Reply:
For the most part, I'm pretty hesitant to add unit tests to `SmartSnippet` just yet, since it may get changed radically. However, the DOM structure of it may not change that much, so I'll add a few unit tests.